delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/CompositeViews.scala

```diff
@@ -0,0 +1,122 @@
+package ch.epfl.bluebrain.nexus.delta.plugins.compositeviews
+
+import cats.effect.Clock
+import ch.epfl.bluebrain.nexus.delta.kernel.utils.{IOUtils, UUIDF}
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model.CompositeViewCommand._
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model.CompositeViewEvent._
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model.CompositeViewRejection._
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model.CompositeViewState.{Current, Initial}
+import ch.epfl.bluebrain.nexus.delta.plugins.compositeviews.model._
+import monix.bio.{IO, UIO}
+
+object CompositeViews {
+
+  type ValidateProjection = CompositeViewProjection => IO[CompositeViewRejection, Unit]
+  type ValidateSource     = CompositeViewSource => IO[CompositeViewRejection, Unit]
+
+  private[compositeviews] def next(
+      state: CompositeViewState,
+      event: CompositeViewEvent
+  ): CompositeViewState = {
+    // format: off
+    def created(e: CompositeViewCreated): CompositeViewState = state match {
+      case Initial     => Current(e.id, e.project, e.uuid, e.value, e.source, Map.empty, e.rev, deprecated = false,  e.instant, e.subject, e.instant, e.subject)
+      case s: Current  => s
+    }
+
+    def updated(e: CompositeViewUpdated): CompositeViewState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, value = e.value, source = e.source, updatedAt = e.instant, updatedBy = e.subject)
+    }
+
+    def tagAdded(e: CompositeViewTagAdded): CompositeViewState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, tags = s.tags + (e.tag -> e.targetRev), updatedAt = e.instant, updatedBy = e.subject)
+    }
+    // format: on
+
+    def deprecated(e: CompositeViewDeprecated): CompositeViewState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, deprecated = true, updatedAt = e.instant, updatedBy = e.subject)
+    }
+
+    event match {
+      case e: CompositeViewCreated    => created(e)
+      case e: CompositeViewUpdated    => updated(e)
+      case e: CompositeViewTagAdded   => tagAdded(e)
+      case e: CompositeViewDeprecated => deprecated(e)
+    }
+  }
+
+  private[compositeviews] def evaluate(validateSource: ValidateSource, validateProjection: ValidateProjection)(
+      state: CompositeViewState,
+      cmd: CompositeViewCommand
+  )(implicit
+      clock: Clock[UIO],
+      uuidF: UUIDF
+  ): IO[CompositeViewRejection, CompositeViewEvent] = {
+
+    def validate(value: CompositeViewValue): IO[CompositeViewRejection, Unit] = for {
+      _ <- IO.parTraverseUnordered(value.sources)(validateSource).void
```

Comment:
It might be counter productive to do it in paralel. Imagine you have many sources with domains that do not exist. It will try to resolve all of them (each of them having to wait until a TCP timeout occurs) and then it will reject. I would rather do it sequentially short circuiting on the first failure you find.